gufe/protocols/protocoldag.py

```diff
@@ -141,9 +141,15 @@ def result_to_unit(self, protocol_unit_result: ProtocolUnitResult):
             raise KeyError("No such `protocol_unit_result` present")
 
     def ok(self) -> bool:
-        # ensure that for every protocol unit, there is definitely a
-        # corresponding result
-        return set(self._protocol_units) == set(self._unit_result_mapping.keys())
+        # ensure there are no failed untis and that for every protocol unit,
+        # there is definitely a corresponding result
+        return (len(self.protocol_unit_failures) == 0 and
+                set(self._protocol_units) == set(self._unit_result_mapping.keys()))
```

Comment:
if we're having Fails as first class, then not having failures isn't a requirement.  e.g I have a DAG with A, B, C units, I run with a scheduler that allows 3 retries, my results might be `[A_ok, B_fail, B_ok, C_ok]`, and that should be `ok() -> True`.  This is what the current `ok()` method does, checks that for every Unit there is at least one success

If we want to have it work this way, we need somewhere to store the Failures

Reply:
Looking at the source we can't currently hold multiple results for a given Unit, but we should imho (and it's easy enough)

Reply:
Oh, the current implementation doesn't actually check for at least one success, but at least one result/failure for each unit. So for a `ProtocolDAG` that fails on only the last unit, it was yielding `ok() -> True`. This is a legit bug that I'll fix here.

Reply:
Do we want to merge #97 and maybe build on the convenience property there?